zarr/_storage/store.py

```diff
@@ -129,6 +131,32 @@ def _ensure_store(store: Any):
             f"wrap it in Zarr.storage.KVStore. Got {store}"
         )
 
+    def getitems(self, keys: Sequence[str], meta_array: NDArrayLike) -> Mapping[str, Any]:
+        """Retrieve data from multiple keys.
+
+        Parameters
+        ----------
+        keys : Iterable[str]
+            The keys to retrieve
+        meta_array : array-like
+            An array instance to use for determining the output type. For now, this is
+            only a hint and can be ignore by the implementation, in which case the type
+            of the output is the same as calling __getitem__() for each key in keys.
+
+        Returns
+        -------
+        Mapping
+            A collection mapping the input keys to their results.
+
+        Developer Notes
+        ---------------
+        This default implementation use __getitem__() to read each key sequential and
+        ignores the meta_array argument. Overwrite this method to implement concurrent
+        reads of multiple keys and/or to utilize the meta_array argument.
+        """
+
+        return {k: self[k] for k in keys if k in self}
```

Comment:
Know fsspec-based storage layers are using a `dict` return value, but am wondering if we should be doing something different (like returning an iterable). Asking since this would make the read blocking vs. a bit more lazy. The latter can be useful when working with operations that take a bit longer (like reading from the cloud or parallelizing several reads)

cc @martindurant (who also may have thoughts)

Reply:
The point of it being blocking, is that  many keys may be being fetched concurrently. If you make it an iterator, you lose that, unless you have something that can wait on an async iterator, which in turn has first-completed working.

Reply:
I agree with both of you.
The API specifies a return type of `Mapping[str, Any]` thus it is possible to return a _lazy_ mapping that only reads `self[k]` when accessed. But let's do that in a follow up PR?

Reply:
Understood, and will be interested in how that looks.

By the way, I mentioned elsewhere the possibility of passing key-specific metadata to the get function; that would happen in this same signature. I wonder if you have any use for an array where only *some* of it is destined for the GPU (perhaps because that data already exists there and doesn't need loading at all).

Reply:
Do you have an example of what they would look like, @martindurant?

Reply:
``` 
def getitems(self, keys: Sequence[str], meta_array: NDArrayLike, key_contexts: Mapping[str, Mapping]) -> Mapping[str, Any]:
```
?

Or did I misunderstand the question here?

Reply:
I don't know of such a use case. If it is a rare case, it might be sufficient to just divide the read into multiple calls to `getitems()` -- one for each return type.

Reply:
No, please don't - we would loose concurrency. Also, that would need getitems() to change instead, so no win. I will try to find the issue where I raise this.

Reply:
Like https://github.com/zarr-developers/zarr-python/pull/1096#issuecomment-1226045840

Reply:
In summary, we have a few different things that may want to pass to the codec/storage chain, each for a specialised case. Examples
- array type we want to end up with
- load many pieces concurrently
- locally-dependent codec parameters (even different compression per chunk)
- partial read of a chunk
- (perhaps) return a lazily or async evaluated thing
- probably others

Partial read is already implemented specifically for blosc, and it took a lot of hackery to fit it in the current code. Here we are adding another special case by peppering kwargs in the code. I am suggesting that it might make more sense to combine these pieces of specialised and optional information into some "context" which can be passed to both codecs and storage. 

Actually, this raises a confusion of mine with this PR: the storage layer never returns an array, it returns a buffer. Is this not the case for kivi or whatever this is directed at?

Reply:
I think a `contexts` is a very good idea. I will be happy to update this PR to implement a more general context approach. 

> Actually, this raises a confusion of mine with this PR: the storage layer never returns an array, it returns a buffer.

Good point, it should probably be a buffer-like type instead.

Reply:
> it should probably be a buffer-like type instead.

So GPU the flow would be: 
- tell the storage layer to produce a GPU buffer
- tell any codec to operate on the GPU and produce GPU buffers (decompression ) or array pieces (filters)
- tell zarr to use these pieces to fill a GPU array

Is it enough to have `memory="gpu"` or similar for all the stages? Will there be cases where more detail about the specific class is required?

Reply:
Is the contexts something that is store specific, Martin? If so, wonder if the particular stores could track that information and one could just use [`collections.ChainMap`]( https://docs.python.org/3/library/collections.html#collections.ChainMap ) to combine them all into one store that has this info under-the-hood.

Reply:
No, the context is something specific to each key, which a store (or codec) may choose to do something with, similar to `block_info` in a dask array.

Reply:
Ok. Still trying to wrap my head around this idea (so apologies if this suggestion is naive). Would it help if supplied an `out`/`outs`/`outputs`/etc. argument that (optionally) provided the arrays to write into?

Reply:
If the type of the output was the only thing we cared about, that would do; but I am trying to motivate something more general. Interestingly, codecs do allow passing a buffer to write into, so it be reasonable to do the same for the storage layer, except that we won't usually know how big the buffer should be.

Reply:
As `getitems` is called by index operations, they should know how big the resulting array will be (and may even have that array provided). So it should be possible to be fairly efficient with memory, which seems important alongside the ability to fill up those pieces asynchronously behind `getitems`.

Could you please share a bit more about the use case you have in mind?

Reply:
I listed some thoughts above, but two immediate concerns are:
- partial read of uncompressed or block-compressed chunks (includes the possibility of subchunks and other sharding style access)
- a codec whose behaviour is dependent on location in the array, e.g., having a scale factor/offset that is a function of the Z dimension of a 3d array

and then the concern immediately addressed by this PR: picking the array class we want to create.

Reply:
With partial reads/writes, I think we should handle this as part of the sharding work either in PR ( https://github.com/zarr-developers/zarr-python/pull/1111 ) or PR ( https://github.com/zarr-developers/zarr-python/pull/1096 ). It is possible we want a different method (like `get_partial_items` and `set_partial_items`), but am not sure about that. Would want Jonathan's input since he has spent more time thinking about sharding APIs.

With location dependent codecs, know this has come up in a few meetings, but Idk if we have created an issue (am not finding one atm). Would suggest we start with a new issue and scope that out. It will most likely involve more changes than just the API here as Numcodecs would need to become aware of this somehow. Also not sure whether it interacts with sharding. So a fair bit to unpack here.

Reply:
One quick thought on the signature: while introducing the `dimension_separator` element, I often had the feeling that my life would have been simplified if there were a (extensible) `Key` object rather than just a string. I wonder if this is a case where we might want to support `Union[str, Key]` rather than introducing a separate `contexts: Mapping[str, Mapping]`

Reply:
I am ambivalent about a Key object, but the above suggestion is almost exactly what I had said first. Whether the contexts should be a mapping or a sequence is not important to me. For now, the context will be empty or include the desired class, and we can agree on things that can be added into it.

Reply:
As [noted above]( https://github.com/zarr-developers/zarr-python/pull/1131#discussion_r990402879 ), think this discussion gets hairy enough it should be broken out into an issue (likely two) and discussed separately.

Reply:
> As [noted above](https://github.com/zarr-developers/zarr-python/pull/1131#discussion_r990402879), think this discussion gets hairy enough it should be broken out into an issue (likely two) and discussed separately.

If we want to go the `contexts` route, we don't need a `meta_array` argument

Reply:
Does this also imply that `getitem()` should have a `contexts` parameter?

Reply:
> Does this also imply that `getitem()` should have a `contexts` parameter?

Good point, yes `getitem()` should also take a  `contexts` parameter